sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java

```diff
@@ -147,7 +150,24 @@ public VectorizedColumnReader(
     DictionaryPage dictionaryPage = pageReader.readDictionaryPage();
     if (dictionaryPage != null) {
       try {
-        this.dictionary = dictionaryPage.getEncoding().initDictionary(descriptor, dictionaryPage);
+        PrimitiveType primitiveType = descriptor.getPrimitiveType();
+        if (primitiveType.getOriginalType() == OriginalType.DECIMAL &&
+            primitiveType.getDecimalMetadata().getPrecision() <= Decimal.MAX_INT_DIGITS() &&
+            primitiveType.getPrimitiveTypeName() == INT64) {
+          // We need to make sure that we initialize the right type for the dictionary otherwise
+          // Encoding#initDictionary will initialize it to PrimitiveTypeName.INT64 and
+          // WritableColumnVector will throw an exception when trying to decode to an Int when the
+          // dictionary is in fact initialized as Long
+          PrimitiveType adjustedType = new PrimitiveType(primitiveType.getRepetition(),
+              INT32, primitiveType.getTypeLength(),
+              primitiveType.getName(), primitiveType.getOriginalType(),
+              primitiveType.getDecimalMetadata(), primitiveType.getId());
+          ColumnDescriptor desc = new ColumnDescriptor(descriptor.getPath(),
+              adjustedType, descriptor.getMaxRepetitionLevel(), descriptor.getMaxDefinitionLevel());
+          this.dictionary = dictionaryPage.getEncoding().initDictionary(desc, dictionaryPage);
```

Comment:
@razajafri I see that here we initialize the dictionary with an adjusted `ColumnDescriptor`. I'm curious about where the INT64 -> INT32 happens. Inside the parquet MR library?

Reply:
Thanks @cloud-fan. You have helped me uncover a bug in my code.

The dictionary is being written as INT64 so just by reading it as an INT32 is not the right solution. I will have to improve my test so this is tested properly. The solution will be to read the dictionary as Long and cast it instead of decoding it directly to Int.

Reply:
@cloud-fan I couldn't find a way to do this any other way but to catch an exception if reading trying to get an Int from the dictionary first, if it fails then we try to read it as a long and cast it. 

Please let me know what you think

Reply:
After talking to @revans2, I have updated the PR with the fix proposed by him. 

I agree that this is a much better change. Please take a look

Reply:
After talking to @revans2, I have updated the PR with the fix proposed by him. 

I agree that this is a much better change. Please take a look